Changes to provide multiple downsampling strategies.

tfenne opened the pull request:
[deleted]

tfenne commented:
@yfarjoun Can I ask you to take a look at this?  @nh13 has already over-the-shoulder reviewed.

tfenne commented:
A couple of notes:
1) I have a change for Picard coming once this PR is merged that incorporates this change and exposes some other options in DownsampleSam.
2) My goal here is to make the downsampling procedure more accurate, while still requiring only a single pass through the file and not requiring significantly more memory *unless* you want super-high accuracy.

It achieves this by reading in batches of reads (if you want accuracy +/- 1/1000 then in reads in records until it has seen 1000 novel _templates_), and then keeping `PROBABILITY * batch_size` records.  It rounds up or down depending on whether it is currently under or over it's goal.

yfarjoun commented:
can you explain what you mean by "more accurate downsampling"? In what way
is the current iterator not accurate?

On Fri, Nov 6, 2015 at 3:55 PM, Tim Fennell <notifications@github.com>
wrote:

> A couple of notes:
> 1) I have a change for Picard coming once this PR is merged that
> incorporates this change and exposes some other options in DownsampleSam.
> 2) My goal here is to make the downsampling procedure more accurate, while
> still requiring only a single pass through the file and not requiring
> significantly more memory *unless* you want super-high accuracy.
>
> It achieves this by reading in batches of reads (if you want accuracy +/-
> 1/1000 then in reads in records until it has seen 1000 novel *templates*),
> and then keeping PROBABILITY * batch_size records. It rounds up or down
> depending on whether it is currently under or over it's goal.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/372#issuecomment-154532041>.
>

tfenne commented:
@yfarjoun Sorry - yes should have explained that.  The current version of the iterator simply makes it so that every template has `PROBABILITY` of being retained, but makes no effort beyond that to ensure that on average input * `PROBABILITY` reads are emitted.  When downsampling to very large numbers (several million reads) you get a fairly accurate output.  But if downsampling to smaller numbers the stochasticity of the process means that it's very hard to use the iterator to produce an output with a specific number of reads in it.

I.e. if I have a file with 10m reads and want 250k for some reason, I might actually end up with 250k +/- 1% when using `PROBABILITY=0.025`.  The goal of this change is to make the result have `output_reads = input_reads * (probability +/- accuracy)`.

yfarjoun commented:
I take issue with your use of "on average", since it is exactly, on average
that input * PROBABILITY records will be emitted...

You seem to want something better than that. You want a smaller
standard-deviation than the expected input*sqrt(P), am I correct?



On Fri, Nov 6, 2015 at 4:48 PM, Tim Fennell <notifications@github.com>
wrote:

> @yfarjoun <https://github.com/yfarjoun> Sorry - yes should have explained
> that. The current version of the iterator simply makes it so that every
> template has PROBABILITY of being retained, but makes no effort beyond
> that to ensure that on average input * PROBABILITY reads are emitted.
> When downsampling to very large numbers (several million reads) you get a
> fairly accurate output. But if downsampling to smaller numbers the
> stochasticity of the process means that it's very hard to use the iterator
> to produce an output with a specific number of reads in it.
>
> I.e. if I have a file with 10m reads and want 250k for some reason, I
> might actually end up with 250k +/- 1% when using PROBABILITY=0.025. The
> goal of this change is to make the result have output_reads = input_reads
> * (probability +/- accuracy).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/372#issuecomment-154551848>.
>

tfenne commented:
I'm not really looking for a semantic argument :disappointed: Yes, I'm trying to make it so that for any single invocation of DownsampleSam (which uses the iterator) you get a more predictable and closer to the target number of reads emitted.  Many people (me included) use this with the goal of reducing a BAM with N input reads to a BAM with N*PROBABILITY reads.  This PR is to make it so that the different between what is output and N*PROBABILITY is within known bounds and not stochastic.

yfarjoun commented:
sorry I can't help it sometime...:-p

yfarjoun commented on src/java/htsjdk/samtools/DownsamplingIterator.java:
```diff
@@ -123,6 +152,45 @@ private boolean advance() {
+            final SAMRecord rec = this.underlyingIterator.next();
+            recs.add(rec);
+
+            if (!this.allowSecondaryAlignments && rec.getNotPrimaryAlignmentFlag()) continue;
```
getNotPrimaryAlignmentFlag=true,  could also mean supplementary, right? 
what if 
```java    
    this.allowSecondaryAlignments==false;
    this.allowSupplementalAlignments==true;
```
and 
```java
    rec.getNotPrimaryAlignmentFlag()==true && rec.getSupplementaryAlignmentFlag()==true
```

I think that the record will not be included, erroneously.

tfenne replied:
No - there's understandably some confusion around this.  `getNotPrimaryAlignmentFlag()` just returns the boolean/flag bit for primary vs. secondary.  There's another method called something like `isSecondaryOrSupplemental()` that does what you're thinking.

yfarjoun replied:
I see. the sam-spec calls this flag  "secondary alignment", but the htsjdk
code it's getNotPrimaryAlignmentFlag...perhaps it's time to deprecate and
have a different function?

not in this PR, clearly!



On Fri, Nov 6, 2015 at 5:23 PM, Tim Fennell <notifications@github.com>
wrote:
>
> In src/java/htsjdk/samtools/DownsamplingIterator.java:
>
> > +     * downsampling can be performed and the target accuracy
achieved.  Once reads have been buffered,
> > +     * template names are randomly sampled out for discarding until
the desired number of reads have
> > +     * been discarded.
> > +     *
> > +     * @return True if one or more reads have been buffered, false
otherwise
> > +     */
> > +    private boolean bufferNextChunkOfRecords() {
> > +        final int templatesToRead = (int) Math.ceil(1 /
this.targetAccuracy);
> > +        final Set<String> names = new HashSet<String>();
> > +        final List<SAMRecord> recs = new
ArrayList<SAMRecord>(templatesToRead);
> > +
> > +        while (this.underlyingIterator.hasNext() && names.size() <
templatesToRead) {
> > +            final SAMRecord rec = this.underlyingIterator.next();
> > +            recs.add(rec);
> > +
> > +            if (!this.allowSecondaryAlignments &&
rec.getNotPrimaryAlignmentFlag()) continue;
>
> No - there's understandably some confusion around this.
getNotPrimaryAlignmentFlag() just returns the boolean/flag bit for primary
vs. secondary. There's another method called something like
isSecondaryOrSupplemental() that does what you're thinking.
>
> —
> Reply to this email directly or view it on GitHub.

yfarjoun commented on src/java/htsjdk/samtools/DownsamplingIterator.java:
```diff
@@ -123,6 +152,45 @@ private boolean advance() {
+            names.add(rec.getReadName());
+        }
+
+        final int templatesRead   = names.size();
```
extra space

yfarjoun commented on src/java/htsjdk/samtools/DownsamplingIterator.java:
```diff
@@ -74,12 +85,30 @@ public DownsamplingIterator setIncludeNoRefReads(final boolean includeNoRefReads
+     * for 1/accuracy templates, so setting this to extremely small numbers is not advisable.
+     */
+    public DownsamplingIterator setTargetAccuracy(final double accuracy) {
+        if (accuracy >= 1 || accuracy <= 0) throw new IllegalArgumentException("Illegal value. Must be 0 < accuracy < 1");
```
to avoid overflow, accuracy must be greater than 1/MAX_INT (this is also true for other reasons, but one could conceive of enough memory to hold 2x10^9 records in memory, and so perhaps it's worth noting, we've been having overflows in other parts of the code due to similar issues.)

yfarjoun commented on src/java/htsjdk/samtools/DownsamplingIterator.java:
```diff
@@ -123,6 +152,45 @@ private boolean advance() {
+        final int templatesRead   = names.size();
+        final double rawTemplatesToKeep = templatesRead * this.probabilityOfKeeping;
+        final int templatesToKeep = (keptTemplates / (double) totalTemplates < this.probabilityOfKeeping)
+                ? (int) Math.ceil(rawTemplatesToKeep) : (int) Math.floor(rawTemplatesToKeep);
```
the alluded to overflow is here

yfarjoun commented on src/java/htsjdk/samtools/DownsamplingIterator.java:
```diff
@@ -123,6 +152,45 @@ private boolean advance() {
+                ? (int) Math.ceil(rawTemplatesToKeep) : (int) Math.floor(rawTemplatesToKeep);
+        final List<String> tmp    = new ArrayList<String>(names);
+        while (names.size() > templatesToKeep) {
+            names.remove(tmp.get(this.random.nextInt(templatesRead)));
```
This could take a while if you are downsampling a large amount (as you'll be hitting already-removed names). Could you make tmp a linkedlist and also remove the drawn integer position from the list? Otherwise, you could hash the names using murmur or similar and take the first templatesToKeep by hash order.

tfenne replied:
I did think about this, and discussed offline with @nh13.  I think a LinkedList would be disastrous for because just getting to the n'th element requires traversing n links in the list.  So while in theory it should be fewer logical operations, the cost of each operation would be many times higher.

The other option I had considered was to just remove elements from the list, and then convert it to a Set at the end.  But I'm not sure that shuffling all that memory is going to be faster either than just putting up with the repeated hits.  I'm happy to change it if you think this would be better.

yfarjoun replied:
then why not shuffle tmp and then remove the first templatesToKeep?

yfarjoun replied:
rather the first size-templatesToKeep

yfarjoun commented:
Got Testz?

This is quite complicated and it would be nice to have some sanity checks...just check that the number of records emitted is in the ball-park, I realize it would be difficult to test for correctness, but P=1, P=0, and P=0.5 should return reasonably sized bams, yes?

Also, since I think we spotted a bug in the logic of secondary and supplementary reads, could you add a unit test for that?

yfarjoun commented:
@tfenne back to you.

tfenne commented:
@yfarjoun Have attempted to address your comments.  I've left them in separate commits that I will squash once we're ready to merge.

The only thing I didn't do (I think) is test your secondary/supplemental case because reading it again, I disagree with you on the correct result.  I think in the case that all of the following are true:

```java    
    this.allowSecondaryAlignments==false;
    this.allowSupplementalAlignments==true;
    rec.getNotPrimaryAlignmentFlag()==true && rec.getSupplementaryAlignmentFlag()==true
```

that the record should be discarded.  There are two hurdles - it makes it past the supplemental hurdle because, although it is supplemental, those are allowed by the second line.  But it fails the secondary/not-primary hurdle because is *not* primary, i.e. secondary, and those are dis-allowed by the first line.

yfarjoun commented:
@tfenne yes, my example was incorrect, as it was based on the misinterpretation of getNotPrimaryAlignmentFlag.

tfenne commented:
@yfarjoun I made the change you suggested, and then decided to try something else.  I'd appreciate it if you and @nh13 could look at my latest commit on this branch/PR.

I've replaced the use of a Map to store retain/discard decisions with a pair of BitSets (one to record decisions, one to record whether or not a decision has been made yet).  And instead of making a decision per template name, it makes a decision per "key", where key is currently implemented as abs(name.hashCode()).  I think the only real question is whether String.hashCode() is distributed uniformly _enough_.  It makes efforts to be uniform, and in practice it seems to work well enogh, but does reduce accuracy a tiny bit.

On the plus side, it now runs in a fixed amount of memory and runs faster due to not querying a HashMap so frequently.

yfarjoun commented:
I recall we had discussed using this approach a while back, and the
conclusion was that a good hash function to use would be murmur3 (I think).
The problem was finding a java implementation that we could use. I think
that String.hashCode() had too many collisions to be considered a good
hash-code for this purpose.




On Sat, Nov 7, 2015 at 7:55 AM, Tim Fennell <notifications@github.com>
wrote:

> @yfarjoun <https://github.com/yfarjoun> I made the change you suggested,
> and then decided to try something else. I'd appreciate it if you and @nh13
> <https://github.com/nh13> could look at my latest commit on this
> branch/PR.
>
> I've replaced the use of a Map to store retain/discard decisions with a
> pair of BitSets (one to record decisions, one to record whether or not a
> decision has been made yet). And instead of making a decision per template
> name, it makes a decision per "key", where key is currently implemented as
> abs(name.hashCode()). I think the only real question is whether
> String.hashCode() is distributed uniformly *enough*. It makes efforts to
> be uniform, and in practice it seems to work well enogh, but does reduce
> accuracy a tiny bit.
>
> On the plus side, it now runs in a fixed amount of memory and runs faster
> due to not querying a HashMap so frequently.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/372#issuecomment-154698865>.
>

tfenne commented:
I did a little testing with a random real BAM from MiSeq data.  The BAM has 17,443,821 actual templates (i.e. unique read names in it).  If I've done my math right, I'd expect by chance to have draw out 17,373,165 unique hashes between 0 and Integer.MAX_VALUE.  With String.hashCode() I get 17,373,668 unique hashes, with Murmur3_32 I get 17,373,368.  Both are higher than the expected, but String.hashCode() is higher than Murmur3_32.  I'm left wondering why we thought String.hashCode would not be random enough - perhaps it is more susceptible to collisions in certain edge cases?

If we do prefer Murmur3, Google Guava has a nice java implementation of Murmur3_32 (and many other hash functions) all under apache 2 licensing which allows for copying and modification of the source.  And conveniently it looks like the clojure guys have already done the work of extracting it to a standalone class here:
    https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/Murmur3.java

yfarjoun commented:
IIRC, the problem was that if string A and string B collide then A+C and
B+C collide for all strings C.


On Sat, Nov 7, 2015 at 10:11 AM, Tim Fennell <notifications@github.com>
wrote:

> I did a little testing with a random real BAM from MiSeq data. The BAM has
> 17,443,821 actual templates (i.e. unique read names in it). If I've done my
> math right, I'd expect by chance to have draw out 17,373,165 unique hashes
> between 0 and Integer.MAX_VALUE. With String.hashCode() I get 17,373,668
> unique hashes, with Murmur3_32 I get 17,373,368. Both are higher than the
> expected, but String.hashCode() is higher than Murmur3_32. I'm left
> wondering why we thought String.hashCode would not be random enough -
> perhaps it is more susceptible to collisions in certain edge cases?
>
> If we do prefer Murmur3, Google Guava has a nice java implementation of
> Murmur3_32 (and many other hash functions) all under apache 2 licensing
> which allows for copying and modification of the source. And conveniently
> it looks like the clojure guys have already done the work of extracting it
> to a standalone class here:
>
> https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/Murmur3.java
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/372#issuecomment-154713167>.
>

yfarjoun commented:
I was basing the analysis on
http://programmers.stackexchange.com/questions/49550/which-hashing-algorithm-is-best-for-uniqueness-and-speed
and it was murmur2 that won (perhaps 3 is new and even better?)



On Sat, Nov 7, 2015 at 11:57 AM, Yossi Farjoun <farjoun@broadinstitute.org>
wrote:

> IIRC, the problem was that if string A and string B collide then A+C and
> B+C collide for all strings C.
>
>
> On Sat, Nov 7, 2015 at 10:11 AM, Tim Fennell <notifications@github.com>
> wrote:
>
>> I did a little testing with a random real BAM from MiSeq data. The BAM
>> has 17,443,821 actual templates (i.e. unique read names in it). If I've
>> done my math right, I'd expect by chance to have draw out 17,373,165 unique
>> hashes between 0 and Integer.MAX_VALUE. With String.hashCode() I get
>> 17,373,668 unique hashes, with Murmur3_32 I get 17,373,368. Both are higher
>> than the expected, but String.hashCode() is higher than Murmur3_32. I'm
>> left wondering why we thought String.hashCode would not be random enough -
>> perhaps it is more susceptible to collisions in certain edge cases?
>>
>> If we do prefer Murmur3, Google Guava has a nice java implementation of
>> Murmur3_32 (and many other hash functions) all under apache 2 licensing
>> which allows for copying and modification of the source. And conveniently
>> it looks like the clojure guys have already done the work of extracting it
>> to a standalone class here:
>>
>> https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/Murmur3.java
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/samtools/htsjdk/pull/372#issuecomment-154713167>.
>>
>
>

tfenne commented:
Interesting reading @yfarjoun  - thanks!  So without doing similar testing or figuring out which algorithm String.hashCode() maps to, Murmur would appear to be safest.

That leaves the question of whether this is preferable.  I think in general it is *except* for cases where the input approaches 2e9 templates as that will cause some undesirable behaviour as the amount of reads needed to see N _new_ templates in `bufferNextChunkOfRecords()` will expand further and further, causing ever larger batches of reads to be buffered.  For example, a 60X human whole genome with 2x150 reads might have > 6e8 templates or 0,3x the hash space.

yfarjoun commented:
Java String hash uses something similar to DJB2 (but with 31 instead of
33). The post has a plot of DJB2a which uses a XOR instead of +. While this
"should" result in even better hash, you can see in the post that DJB2a is
not nearly as good as murmur2/3.

I support the inclusion of murmur3 into htsjdk, though note that the
cloture class isn't standalone, it needs Util.java.



On Sat, Nov 7, 2015 at 12:50 PM, Tim Fennell <notifications@github.com>
wrote:

> Interesting reading @yfarjoun <https://github.com/yfarjoun> - thanks! So
> without doing similar testing or figuring out which algorithm
> String.hashCode() maps to, Murmur would appear to be safest.
>
> That leaves the question of whether this is preferable. I think in general
> it is *except* for cases where the input approaches 2e9 templates as that
> will cause some undesirable behaviour as the amount of reads needed to see
> N *new* templates in bufferNextChunkOfRecords() will expand further and
> further, causing ever larger batches of reads to be buffered. For example,
> a 60X human whole genome with 2x150 reads might have > 6e8 templates or
> 0,3x the hash space.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/372#issuecomment-154729927>.
>

yfarjoun commented:
Given that the hashcode can be a long not an int, we could also allow more
than 32 bits for the hashcode. From the bamindex one could easily find the
total number of reads, take some multiple of that to inform the hash-space
you need and murmur will provide...right? the memory usage will increase
for large bams, but that was already the case with the map

The only problem is that the cloture implementation provides the 32 bit
version, not the 64 or 128 bit versions that exist in Guava, and if I
understand correctly we do not want to include Guava in HTSJDK..





On Sat, Nov 7, 2015 at 5:14 PM, Yossi Farjoun <farjoun@broadinstitute.org>
wrote:

> Java String hash uses something similar to DJB2 (but with 31 instead of
> 33). The post has a plot of DJB2a which uses a XOR instead of +. While this
> "should" result in even better hash, you can see in the post that DJB2a is
> not nearly as good as murmur2/3.
>
> I support the inclusion of murmur3 into htsjdk, though note that the
> cloture class isn't standalone, it needs Util.java.
>
>
>
> On Sat, Nov 7, 2015 at 12:50 PM, Tim Fennell <notifications@github.com>
> wrote:
>
>> Interesting reading @yfarjoun <https://github.com/yfarjoun> - thanks! So
>> without doing similar testing or figuring out which algorithm
>> String.hashCode() maps to, Murmur would appear to be safest.
>>
>> That leaves the question of whether this is preferable. I think in
>> general it is *except* for cases where the input approaches 2e9
>> templates as that will cause some undesirable behaviour as the amount of
>> reads needed to see N *new* templates in bufferNextChunkOfRecords() will
>> expand further and further, causing ever larger batches of reads to be
>> buffered. For example, a 60X human whole genome with 2x150 reads might have
>> > 6e8 templates or 0,3x the hash space.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/samtools/htsjdk/pull/372#issuecomment-154729927>.
>>
>
>

tfenne commented:
@yfarjoun @nh13 After spending some more time thinking about this I decided that the correct way forward was to provide multiple ways of downsampling to cater to multiple use cases.  The use cases I am trying to cover are:
1) Downsampling very large input files without needs a ton of memory
2) Downsampling moderately sized input files to small numbers of reads with high accuracy ( @yfarjoun I've converged on defining accuracy as the absolute difference between the requested proportion and the emitted proportion)
3) Downsampling very large input files to relatively small numbers of reads at higher accuracy than #1 provides.

I've rebased and squashed since the latest changes are quite a shift in direction.  I'm much happier with the results though. 

I should probably also note that I have decided that it is *not* the responsibility of downsampling iterators to filter reads in any way other than downsampling (i.e. secondary, supplemental, unmapped) and that can easily be achieved by using a `FilteringIterator` under a `DownsamplingIterator`.

yfarjoun commented:
Sounds good. I'll take a look. I agree 100% with leaving the filtering to a
subsequent, or previous filteringiterator.

On Sun, Nov 8, 2015 at 9:46 AM, Tim Fennell <notifications@github.com>
wrote:

> @yfarjoun <https://github.com/yfarjoun> @nh13 <https://github.com/nh13>
> After spending some more time thinking about this I decided that the
> correct way forward was to provide multiple ways of downsampling to cater
> to multiple use cases. The use cases I am trying to cover are:
> 1) Downsampling very large input files without needs a ton of memory
> 2) Downsampling moderately sized input files to small numbers of reads
> with high accuracy ( @yfarjoun <https://github.com/yfarjoun> I've
> converged on defining accuracy as the absolute difference between the
> requested proportion and the emitted proportion)
> 3) Downsampling very large input files to relatively small numbers of
> reads at higher accuracy than #1
> <https://github.com/samtools/htsjdk/pull/1> provides.
>
> I've rebased and squashed since the latest changes are quite a shift in
> direction. I'm much happier with the results though.
>
> I should probably also note that I have decided that it is *not* the
> responsibility of downsampling iterators to filter reads in any way other
> than downsampling (i.e. secondary, supplemental, unmapped) and that can
> easily be achieved by using a FilteringIterator under a
> DownsamplingIterator.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/372#issuecomment-154830781>.
>

nh13 commented on build.xml:
```diff
@@ -157,7 +157,7 @@
         <testng suitename="htsjdk-single-test" classpathref="classpath" outputdir="${test.output}"
                 verbose="${testng.verbosity}">
-            <jvmarg line="-Xmx512M ${debug.jvm.args}"/>
+            <jvmarg line="-Xmx2g ${debug.jvm.args}"/>
```
be careful here, as Travis CI only has 3Gb of memory.

tfenne replied:
Thanks for catching this!  I had changed it when I was experimenting with using a pair of BitSets that sucked up quite a bit of memory and forgot to change it back.

nh13 commented on src/java/htsjdk/samtools/ConstantMemoryDownsamplingIterator.java:
```diff
@@ -0,0 +1,87 @@
+ * are emitted, records whose hash value is above the limit are discarded.
+ *
+ * Does not make any attempt to be accurate (have actual proportion == requested proportion) beyond what would
+ * be expected for a random process and so may become quite in accurate when downsampling to small numbers of
```
`in accurate` -> `inaccurate`

nh13 commented on src/java/htsjdk/samtools/DownsamplingIteratorFactory.java:
```diff
@@ -0,0 +1,117 @@
+            "to working in fixed memory this strategy is good for large inputs, and due to the stochastic nature the accuracy of this strategy " +
+            "is highest with a high number of output records, and diminishes at low output volumes.";
+
+    public static final String CHAINED_DESCRPTION =
```
`CHAINED_DESCRPTION` -> `CHAINED_DESCRIPTION`

nh13 commented on src/java/htsjdk/samtools/HighAccuracyDownsamplingIterator.java:
```diff
@@ -0,0 +1,196 @@
+            final SAMRecord rec = this.bufferedRecords.next();
+            final String key = rec.getReadName();
+            final Boolean previous = decisions.get(key);
+            final boolean keeper;
```
I can't stop thinking about a goalie, how about renaming it to `keepRecord`?

nh13 commented on src/java/htsjdk/samtools/HighAccuracyDownsamplingIterator.java:
```diff
@@ -0,0 +1,196 @@
+
+    /**
+     * Buffers reads until either the end of the file is reached or enough reads have been buffered that
+     * downsampling can be performed and the target accuracy achieved.  Once reads have been buffered,
```
is achieved? can be achieved?

nh13 commented on src/java/htsjdk/samtools/HighAccuracyDownsamplingIterator.java:
```diff
@@ -0,0 +1,196 @@
+    protected int calculateTemplatesToKeep(final int templatesRead, final double proportion) {
+        final double rawTemplatesToKeep = templatesRead * proportion;
+        return (keptTemplates / (double) totalTemplates < proportion)
+                ? (int) Math.ceil(rawTemplatesToKeep) : (int) Math.floor(rawTemplatesToKeep);
```
cute

nh13 commented:
@tfenne nothing to add here but a few typos.  I am excited to take it for a spin.

tfenne commented:
Thanks @nh13.  Just pushed a commit to address your comments.

@yfarjoun any objects to merging once the tests pass?  I'll get the Picard PR queued up so that it can be reviewed merged quickly.  The changes there are small, but necessary.

yfarjoun commented:
I still need a bit more time to get through the CHAINED stuff. I read the
other strategies. sorry for the delay.

On Mon, Nov 9, 2015 at 6:37 AM, Tim Fennell <notifications@github.com>
wrote:

> Thanks @nh13 <https://github.com/nh13>. Just pushed a commit to address
> your comments.
>
> @yfarjoun <https://github.com/yfarjoun> any objects to merging once the
> tests pass? I'll get the Picard PR queued up so that it can be reviewed
> merged quickly. The changes there are small, but necessary.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/372#issuecomment-155038360>.
>

tfenne commented:
No problem @yfarjoun.  Appreciate the in depth review!

tfenne commented:
@yfarjoun any chance you can take a look at the CHAINED stuff today so we can wrap this review?

yfarjoun commented:
"We're sorry, we're experiencing an unusually large volume of pull
requests. All of our operators are currently busy or on vacation. Your PR
is important to us and will be reviewed as soon as possible."

:-)

I'll try.



On Tue, Nov 10, 2015 at 8:24 AM, Tim Fennell <notifications@github.com>
wrote:

> @yfarjoun <https://github.com/yfarjoun> any chance you can take a look at
> the CHAINED stuff today so we can wrap this review?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/372#issuecomment-155417994>.
>

nh13 commented:
All pull requests left unattended will be subject to immediate merging.  Please remove all :scissors: and :mans_shoe: from your luggage before proceeding through security.

yfarjoun commented:
strange test failure.

yfarjoun commented on src/java/htsjdk/samtools/ChainedDownsamplingIterator.java:
```diff
@@ -0,0 +1,79 @@
+     * Constructs a chained iterator that will read from the provided iterator and attempt to downsampling to the provided proportion.
+     */
+    ChainedDownsamplingIterator(final Iterator<SAMRecord> iterator, final double proportion, final int seed) {
+        super(new ConstantMemoryDownsamplingIterator(iterator, Math.min(1, proportion + 0.01), seed), proportion, seed);
```
presumably, the 1% extra is so that you will "definitely" have the amount you need, right? in that case I'd use 5sd which for the [Bernoulli](https://en.wikipedia.org/wiki/Bernoulli_distribution) is \sqrt(p(1-p)). so I'd recommend using p + 5*sqrt(p*(1-p)) instead of p+0.01. 

Good thing you allow 1 as a valid value for p, :smiley_cat:

tfenne replied:
Thanks - will change.  I was kind of hoping you'd tell me the right math to use here ;)  Incidentally, why 5sd and not 4 or 6?

tfenne replied:
Wait a minute - that doesn't work :/  E.g. for p=0.8, you get p(1-p) = 0.16.  sqrt(0.16) = 0.4, and 5*0.4 = 2.

tfenne replied:
Thinking more about this, I don't think I can actually estimate a "correct" value to add to P here without knowing in advance the number of reads I will be selecting from.  I.e. I need to know both P and N to calculate standard error or a CI, and I won't know N until I've traversed the data.

yfarjoun replied:
right...we need to take the sqrt of N as well...oops.

tfenne replied:
Ok - how's this for a compromise.  I calculate using a fixed value of N, say 250,000 - and add to the documentation that it is accurate to x% when aiming for >= N reads?

yfarjoun commented on src/java/htsjdk/samtools/ChainedDownsamplingIterator.java:
```diff
@@ -0,0 +1,79 @@
+
+        // Deal with the fact that the iterator will advance and discard some reads at construction
+        final long discarded = ((ConstantMemoryDownsamplingIterator) getUnderlyingIterator()).getDiscardedCount();
+        for (long i=0; i<discarded; ++i) recordDiscardedRecord(null);
```
sigh-spaces?

tfenne replied:
A lifetime's habit is hard to break.  Also, does anyone think that
`for (long i = 0; i < discarded; ++i)`
is actually more readable than
`for (long i=0; i<discarded; ++i)`
because I actually find it harder to read.

nh13 replied:
Sadly, I find it harder to read `for (long i=0; i<discarded; ++i)`

tfenne commented:
@yfarjoun Did the math with an assumed N=50,000 and added to the documentation.  And added your spaces :sob:   Good to go when the tests pass?

tfenne commented:
Woo - and no spurious test failure this time :grin:

yfarjoun commented on src/java/htsjdk/samtools/ChainedDownsamplingIterator.java:
```diff
@@ -0,0 +1,92 @@
+        // Record all the discarded records to keep the overall statistics accurate, but do it after
+        // the call to super() so it doesn't affect the proportion calculation.
+        for (long i = 0; i < iter.getDiscardedCount(); ++i) {
+            recordDiscardedRecord(null);
```
it might make sense to provide a recordDiscarded(int) instead of this silly loop. (and above)

yfarjoun commented:
small comment and then :+1:
